mautic/mautic
# Zoho integration improvements

closes https://github.com/mautic/mautic/issues/9281

**Please be sure you are submitting this against the _staging_ branch.**

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | x
| New feature? | 
| Automated tests included? |
| Related user documentation PR URL | 
| Related developer documentation PR URL | 
| Issues addressed (#s or URLs) | https://github.com/mautic/mautic/issues/9281
| BC breaks? | 
| Deprecations? | 

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:

We build a lot of improvements to  Zoho integration

Bug fixes:
- fix multi select field options on both side
- fix sync company info to Mautic from Leads and Contacts
- pretend sync If contacts already exits in integration 

Improvements: 
-  add new option - Sync blank values from/to integration


# Steps to reproduce: 

1 Create a Multiple Select Custom Field on Mautic
2 Create A Multiple Select Custom Field on Zoho (Settings > Customization > Modules and Fields > Contacts - Select 'Fields' tab and click 'Create & Edit Fields' then drag a multi-select field into the form area. Note also: "If the ZohoCRM field is a pick list, **be sure the list values of Mautic's field match those of the field in ZohoCRM**" [source](https://docs.mautic.org/en/plugins/zoho-crm#test-the-plugin)
3 Mapped those 2 field in the Zoho plugin
4 Create a contact in Mautic, choose 2 choices in the select field
5 Synchronize and see that the 2 choices are in only one in Zoho
6, Map Account Name, Company field in Mautic for both tab (see example of mapping bellow)
6, Create contact on Zoho and add company to this contact
7. Try sync to Mautic. No company there, just on sync run second time
8. Create lead with company name
9. Sync no company name synced
10. Try update values with empty value - for example match Mautic address1 with  Mailing Street on Zoho. 
11. Empty values are skipped. New options added to integration allow sync also empty values 
12. Try to find contact in Zoho Contacts with modified date older like 15 minutes
13. Create same contact in Mautic - email should be same
14. Then try sync Zoho
15. Newly created contact in Mautic is created in Zoho Leads, even thought already exist in Zoho Contacts. We expect contact in Mautic will find already exists contact in Zoho Contacts


Example of mapping
![image](https://user-images.githubusercontent.com/462477/79763830-dbec9e00-8324-11ea-8634-25e915bf2bf6.png)


#### Steps to test this PR:
1. Load up [this PR](https://mautibox.com)
2. Repeat all steps, see multi select in right way also company sync in right way


<a href="https://gitpod.io/#https://github.com/mautic/mautic/pull/7867"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>
